packages/elastic-safer-lodash-set/fp/set.js

```diff
@@ -0,0 +1,9 @@
+/*
+ * Elasticsearch B.V licenses this file to you under the MIT License.
+ * See `packages/elastic-safer-lodash-set/LICENSE` more information.
+ */
+
+const set = require('../lodash/set');
+const { cloneDeep, curry } = require('lodash');
+
+module.exports = curry((b, c, a) => set(cloneDeep(a), b, c));
```

Comment:
I actually didn't look at the source code for how `fp.set` was generated (it's a bit hard to follow as you also have discovered). But when writing the tests for it, I noticed that it behaved differently than `set` (besides simply the argument ordering) in that it would not mutate the input. This is also in line with my understanding of functional programming where you always try to avoid side-effects. From what I can see, it's doing a deep clone somewhere in the internals.

If I output the generated source code from the original `fp.set`, it looks like this:

```js
function() {
/* [wrapped with _.ary, _.curry, & _.rearg] */
    var length = arguments.length;
    if (!length) {
      return;
    }
    var args = Array(length);
    while (length--) {
      args[length] = arguments[length];
    }
    var result = args[0] = cloner.apply(undefined, args);
    func.apply(undefined, args);
    return result;
  }
```

As you can see it's calling `cloner` with arguments and overwriting `args[0]` with the result. I assume this is what's doing the cloning.

Reply:
I actually thought that they use shallow `clone` defined [here](https://github.com/lodash/lodash/blob/4.17.15-npm/fp/_util.js#L4), but I'd not bet my last dollar on this :slightly_smiling_face: Just to be on the safe side can we check that we get exactly same result for the "deep" object when we use our fp `set` and the one provided by lodash/fp? Something like:
```js
const obj = {
  a: { b: 1 },
  c: { d: 2 },
};
// and if we set `a.b`, would `c` be cloned as well in our and lodash implementations?
```

Reply:
You're spot on! In this case `c` would not be cloned in the `lodash/fp` implementation, but will in ours because of the call to `deepClone`. I'll fix it...

Reply:
Fixed by simply doing exactly what Lodash does internally... I can do that now where `lodash` is a peerDependency